include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -157,28 +160,94 @@ struct FFT<2,true, Scalar, device_capabilities>
 
 // ---------------------------- Below pending --------------------------------------------------
 
-/*
-
-// then define 4,8,16 in terms of calling the FFT<2> and doing the special radix steps before/after
-template<uint16_t K, bool Inverse, class device_capabilities>
-struct FFT
+// Forward FFT
+template<uint32_t K, typename Scalar, class device_capabilities>
+struct FFT<K, false, Scalar, device_capabilities>
 {
-    template<typename Accessor, typename ShaderMemoryAccessor>
-    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(ShaderMemoryAccessor) sharedmemAccessor)
+    template<typename Accessor, typename SharedMemoryAccessor>
+    static enable_if_t< (mpl::is_pot_v<K> && K > 2), void > __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
     {
-        if (!Inverse)
+        for (uint32_t stride = (K >> 1) * _NBL_HLSL_WORKGROUP_SIZE_; stride > _NBL_HLSL_WORKGROUP_SIZE_; stride >>= 1)
         {
-           ... special steps ...
+            //[unroll(K/2)]
+            for (uint32_t virtualThreadID = SubgroupContiguousIndex(); virtualThreadID < (K >> 1) * _NBL_HLSL_WORKGROUP_SIZE_; virtualThreadID += _NBL_HLSL_WORKGROUP_SIZE_)
+            {
+                const uint32_t loIx = ((virtualThreadID & (~(stride - 1))) << 1) | (virtualThreadID & (stride - 1));
+                const uint32_t hiIx = loIx | stride;
+                
+                complex_t<Scalar> lo = accessor.get(loIx);
+                complex_t<Scalar> hi = accessor.get(hiIx);
+                
+                hlsl::fft::DIF<Scalar>::radix2(hlsl::fft::twiddle<false,Scalar>(virtualThreadID & (stride - 1), stride),lo,hi);
+                
+                accessor.set(loIx, lo);
+                accessor.set(hiIx, hi);
+            }
+            accessor.memoryBarrier(); // no execution barrier just making sure writes propagate to accessor
```

Comment:
let me just do one quick verification. We need to check that the same invocation still is responsible for computing all elements of a butterfly (because of no execution barrier & we want to be able to use a "register accessor").

Let's say you have a 512 wg, and 8196 fft size.

Reply:
First stride is 4196, so invocation i touches: i,i+4096, i+512,i+4618, etc. at offsets this is 0,4096,512,4618,1024,5120,...

Then stride 2048, so `i` touches at offsets: 0,2048,512,2560,1024,3084,1536,3596, 4096,6144,4618,6656,....

Ok I'm seeing a pattern, we go through all multiples of `WORKGROUP_SIZE` one way or another, just in different orders. The register optimization will work.

Mark as resolved as soon as you read.